Add documents for kuberc (#50913)
kubernetes/website: content/en/docs/reference/kubectl/kuberc.md

---
title: Kubectl user preferences (kuberc)
content_type: concept
weight: 70
---

> 

Reviewer: ```suggestion

{{< feature-state state="alpha" for_k8s_version="1.33 >}}

```

> kuberc allows users to define preferences such as default options, command aliases, and more, using a dedicated configuration file.

Reviewer: ```suggestion
A Kubernetes `kuberc` configuration file allows you to define preferences for kubectl, such as default options and command aliases.
Unlike the kubeconfig file, a `kuberc` configuration file does **not** contain cluster
details, usernames or passwords.
```

> Default location of this configuration file is `$HOME/.kube/kuberc` and it can be customized via `--kuberc` flag similar to kubeconfig.  

Reviewer: ```suggestion

The default location of this configuration file is `$HOME/.kube/kuberc`. You can instruct `kubectl` to look at a custom path for this configuration using the `--kuberc` command line argument.
```

> ## Aliases

Reviewer: I would use lowercase `aliases` here, for clarity.

> Define custom shortcuts for kubectl commands with predefined flags and arguments.

Reviewer: ```suggestion
Within a `kuberc` configuration, _aliases_ allow you to define custom shortcuts
for kubectl commands, optionally with preset command line arguments.
```

> ### Name

Reviewer: I would use lowercase `name` here, for clarity.

Alias name must not collide with the built-in commands. 

> ### Command

Reviewer: I would use lowercase `command` here, for clarity.

Specify the underlying built-in command that your alias will execute. This includes support for subcommands like `create role`.

> ### Flags

Reviewer: I would use lowercase `flags` here, for clarity.

> Customize the defaults values of flags. If you explicitly specify a flag on your terminal, that value will always take precedence over the default one defined in kuberc. 

Reviewer: ```suggestion
Specify default values for command line arguments (which the kuberc format terms _flags_). If you explicitly specify a command line argument when you run kubectl, the value you provide takes precedence over the default one defined in kuberc.
```

> **Example:**

Reviewer: nit: this should be a heading.

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preference
aliases:
- name: getn
  command: get
  flags:
   - name: output
     default: json
```

With this alias, running `kubectl getn pods` will default JSON output. However, if you execute `kubectl getn pods -oyaml`, the output will be in YAML format.

> ### PrependArgs

Reviewer: nit: I'd write `prependArgs`

Insert arbitrary arguments immediately after the kubectl command and its subcommand (if any).

**Example:**

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preference
aliases:
  - name: getn
    command: get
    prependArgs:
      - namespace
    flags:
      - name: output
        default: json
```

> `kubectl getn test-ns` will be translated to `kubectl get namespace test-ns`.

Reviewer: Is this true?
```suggestion
`kubectl getn test-ns` will be translated to `kubectl get namespace test-ns --output json`.
```

> ### AppendArgs

Reviewer: nit: I'd write `appendArgs`

Append arbitrary arguments to the end of the kubectl command.

> **Example:**

Reviewer: nit: this should be a heading

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preference
aliases:
- name: runx
  command: run
  flags:
    - name: image
      default: busybox
    - name: namespace
      default: test-ns
  appendArgs:
    - --
    - custom-arg
```

> `kubectl runx test-pod` will be translated to `kubectl run test-pod -- custom-arg`.

Reviewer: ```suggestion
`kubectl runx test-pod` will be translated to `kubectl run test-pod --namespace test-ns --image busybox -- custom-arg`.
```

## Command Overrides

> Customize the default settings of flags for built-in kubectl commands.

Reviewer: Within a `kuberc` configuration, _command overrides_ let you specify custom values for command line arguments.

> ### Command

Reviewer: I would use lowercase `command` here, for clarity.

Specify the built-in command. This includes support for subcommands like `create role`.

> ### Flags

Reviewer: I would use lowercase `flags` here, for clarity.

> Customize the defaults values of flags. If you explicitly specify a flag on your terminal, explicit value will always take precedence over the default one defined in kuberc.

Reviewer: I think the way we've designed this configuration file format makes it hard to teach, because the names are counter-deductive. Anyway, here's how I'd write it:

```suggestion
Within a `kuberc`, configuration, command line arguments are termed _flags_ (even if they do not represent a boolean type).
You can use `flags` to set the default value of a command line argument.

If you explicitly specify a flag on your terminal, explicit value will always take precedence over the value you defined in kuberc using `overrides`.

{{< note >}}
You cannot use `kuberc` to override the value of a command line argument to take precedence over what the user specifies on the command line. The term `overrides`
in this context refers to specifying a default value that is different from the
compiled-in default value.
{{< /note >}}

```

> **Example:**

Reviewer: nit: This should be a heading.

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preference
overrides:
- command: delete
  flags:
    - name: interactive
      default: "true"
```

With this override, running `kubectl delete pod/test-pod` will default to prompting for confirmation. 
However, `kubectl delete pod/test-pod --interactive=false` will bypass the confirmation.

> As kubectl maintainers, we strongly encourage you to adopt kuberc with the given overrides:

Reviewer: At the project level, we avoid strongly recommending that people adopt alpha features.
```suggestion
The kubectl maintainers encourage you to adopt kuberc with the given defaults:
```

> As kubectl maintainers, we strongly encourage you to adopt kuberc with the given overrides:

Author: Makes sense. Updated.

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preference
overrides:
  - command: apply
    flags:
      - name: server-side
        default: "true"
  - command: delete
    flags:
      - name: interactive
        default: "true"
```

> ## Disable Kuberc

Reviewer: nit: lowercase k (to match the default filename)

To temporarily disable the kuberc functionality, simply export the environment variable `KUBERC` with the value `off`:

```shell
export KUBERC=off

> ```

Reviewer: Given it's an alpha feature, couldn't you just unset something?

Author: I've mentioned disabling via feature gate as well. Please let me know what you think.

content_type: concept
weight: 70
---

{{< feature-state state="alpha" for_k8s_version="1.33" >}}

A Kubernetes `kuberc` configuration file allows you to define preferences for kubectl, such as default options and command aliases.
Unlike the kubeconfig file, a `kuberc` configuration file does **not** contain cluster details, usernames or passwords.

The default location of this configuration file is `$HOME/.kube/kuberc`. 
You can instruct `kubectl` to look at a custom path for this configuration using the `--kuberc` command line argument.  

## aliases

Within a `kuberc` configuration, _aliases_ allow you to define custom shortcuts
for kubectl commands, optionally with preset command line arguments.

### name

Alias name must not collide with the built-in commands. 

### command

Specify the underlying built-in command that your alias will execute. This includes support for subcommands like `create role`.

### flags

Specify default values for command line arguments (which the kuberc format terms _flags_). 
If you explicitly specify a command line argument when you run kubectl, the value you provide takes precedence over the default one defined in kuberc. 

> #### Example:

Author: Updated

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preference
aliases:
- name: getn
  command: get
  flags:
   - name: output
     default: json
```

With this alias, running `kubectl getn pods` will default JSON output. However, if you execute `kubectl getn pods -oyaml`, the output will be in YAML format.

### prependArgs

Insert arbitrary arguments immediately after the kubectl command and its subcommand (if any).

> #### Example:

Reviewer: ```suggestion
#### Example {#prependArgs-example}
```

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preference
aliases:
  - name: getn
    command: get
    prependArgs:
      - namespace
    flags:
      - name: output
        default: json
```

`kubectl getn test-ns` will be translated to `kubectl get namespace test-ns --output json`.

### appendArgs

Append arbitrary arguments to the end of the kubectl command.

#### Example:

```yaml
apiVersion: kubectl.config.k8s.io/v1alpha1
kind: Preference
aliases:
- name: runx
  command: run
  flags:
    - name: image
      default: busybox
    - name: namespace
      default: test-ns
  appendArgs:
    - --
    - custom-arg
```

`kubectl runx test-pod` will be translated to `kubectl run test-pod --namespace test-ns --image busybox -- custom-arg`.

> ## Command Overrides

Reviewer: (nit)
```suggestion
## Command overrides
```

Within a `kuberc` configuration, _command overrides_ let you specify custom values for command line arguments.

### command

Specify the built-in command. This includes support for subcommands like `create role`.

### flags

Within a `kuberc`, configuration, command line arguments are termed _flags_ (even if they do not represent a boolean type).
You can use `flags` to set the default value of a command line argument.

> 	If you explicitly specify a flag on your terminal, explicit value will always take precedence over the value you defined in kuberc using `overrides`.

Reviewer 2: This indentation also causes a slight rendering issue with the page content, but aside from that, the overall changes look good to me!

[Preview Link (here)](https://deploy-preview-50913--kubernetes-io-main-staging.netlify.app/docs/reference/kubectl/kuberc/#flags-1)

<img width="794" alt="pr-preview" src="https://github.com/user-attachments/assets/4233e348-0132-4d0e-a256-874bb50945a4" />
